watertap/core/zero_order_sido_reactive.py

```diff
@@ -0,0 +1,416 @@
+
+    Four additional constraints are added to represent the material balances
+        * water_recovery_equation (indexed by time)
+        * flow_balance (indexed by time)
+        * solute_removal_equation (indexed by time and solute)
+        * solute_treated_equation (indexed by time and solute)
+
+    This method also sets private attributes on the unit model with references
+    to the appropriate initialization and scaling methods to use and to return
+    the inlet volumetric flow rate.
+    """
+    self._has_recovery_removal = True
+    self._initialize = initialize_sidor
+    self._scaling = calculate_scaling_factors_sidor
+
+    # Create state blocks for inlet and outlets
+    tmp_dict = dict(**self.config.property_package_args)
+    tmp_dict["has_phase_equilibrium"] = False
+    tmp_dict["defined_state"] = True
+
+    self.properties_in = self.config.property_package.build_state_block(
+        self.flowsheet().time,
+        doc="Material properties at inlet",
+        default=tmp_dict)
+
+    tmp_dict_2 = dict(**tmp_dict)
+    tmp_dict_2["defined_state"] = False
+
+    self.properties_treated = \
+        self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of treated water",
+            default=tmp_dict_2)
+    self.properties_byproduct = \
+        self.config.property_package.build_state_block(
+            self.flowsheet().time,
+            doc="Material properties of byproduct stream",
+            default=tmp_dict_2)
+
+    # Create Ports
+    self.add_port("inlet", self.properties_in, doc="Inlet port")
+    self.add_port("treated",
+                  self.properties_treated,
+                  doc="Treated water outlet port")
+    self.add_port("byproduct",
+                  self.properties_byproduct,
+                  doc="Byproduct outlet port")
+
+    # Load reaction identifiers from database
+    dbparams = self.config.database.get_unit_operation_parameters(
+        self._tech_type,
+        subtype=self.config.process_subtype)
+    try:
+        rxn_ids = list(dbparams["reactions"].keys())
+    except KeyError:
+        raise KeyError(
+            f"{self.name} - database provided does not contain a list of "
+            f"reactions for this technology.")
+    # Create indexing set for reactions
+    self.reaction_set = Set(initialize=rxn_ids)
+
+    # Add performance variables
+    self.recovery_frac_mass_H2O = Var(
+        self.flowsheet().time,
+        initialize=0.8,
+        domain=NonNegativeReals,
+        units=pyunits.dimensionless,
+        bounds=(1E-8, 1.0000001),
+        doc='Mass recovery fraction of water in the treated stream')
+    self.removal_frac_mass_solute = Var(
+        self.flowsheet().time,
+        self.config.property_package.solute_set,
+        domain=NonNegativeReals,
+        initialize=0.01,
+        units=pyunits.dimensionless,
+        doc='Solute removal fraction on a mass basis')
```

Comment:
I strongly suggest not including water in here for two reasons.

1. It adds the possibility of over-specifying the problem as a user could specify both the removal and recovery of water (or worse yet, specify both of those and leave another removal unfixed and wonder why their problem with 0 degrees of freedom does not solve). Believe me, someone will try to do this. On top of this, think of the additional logic that would need to be implemented to work out which of these to fix when loading parameters from a datafile. What do we do if a user provides values for both recovery and removal?
4. It adds an additional Var and Constraint to the problem that are not really required. We want to keep these models as simple as possible, so I don't see much value in adding an additional Var and Constraint for this.

On the comment about naming, that is primarily semantics. At the moment, "solute" just means anything that is not water, and yes, not everything in there is strictly a solute (it also includes suspended solids, other solvents and potentially entrained gases). However, going with the full component list is not the solution to this. At most, we should use a more inclusive name (but I have no idea what that is). Further, if we do that then we need to go change everything else in the zero-order models too.

Reply:
Just my two cents- while I agree with @TimBartholomew that we don't _need_ to differentiate water from all other components, I think that we don't have time to spare to step back and revise everything that is already in place for WaterTAP3-related models because it would set us back. Instead, I think our time would be better spent by pushing forward and get all models implemented first. So I have to agree with @andrewlee94 on this one. With regard to not all "solutes" necessarily being "solutes", I suppose a more appropriate name would be something like "constituent", but again, since time is limited and altering the name across the board wouldn't change code functionality, I think we can live with what is in place now. Once all of our models are in place and we shift our focus towards refinement, we could go back and refine as needed.